### PR TITLE
Enable greedy sampling

### PR DESCRIPTION
**Pull request by aashiqmuhamed**
This CR enables greedy sampling in model.generate on XLA devices such as Trainium and TPU. This addresses issues such as https://github.com/huggingface/transformers/issues/18661 and https://github.com/huggingface/transformers/issues/12322.

The implementation is inspired by the corresponding Tensorflow generate function in transformers. The CR uses conditional statements to support greedy sampling, and also implements kv-cache functionality that is XLA compatible.

### CODE REVIEW

**Comment by aashiqmuhamed** [comment]
@michaelbenayoun Could you please take a look when you get a chance?

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:170]: So we want the `decoder_attention_mask` to look something like `[[0, 0, 0, .., 1, 1]]`? (batch size = 1 in this example)
Shouldn't the ones come first?

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:190]:
```suggestion
            """Updates the appropriate attention mask -- encoder-decoder models use `decoder_attention_mask`"""
```

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:198]: So we remove the first token, and append a one in the end?
Just making sure I understand.

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:210]:
```suggestion
            """Initializes past_key_values with zeros -- the structure depends on `batch_axis`"""
```

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:212]: To remove?

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:392]: Can be removed?

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:525]:
```suggestion
```

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:953]: Maybe return an error before all of this to notify the user that it is not supported currently

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:75]: You are right, will import from transformers.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:190]: Will make the change.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:206]: Thanks for the suggestion.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:212]: Thanks for catching that.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:216]: Thanks for the suggestion.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:392]: Yes, will remove.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:953]: Simplified the generate function in the revision, to only include greedy sampling.

File: tests/test_generate.py
Comment (aashiqmuhamed) [reply tests/test_generate.py:12]: Thanks for the suggestion.

File: tests/test_generate.py
Comment (aashiqmuhamed) [reply tests/test_generate.py:47]: We are testing sampling for GPT and BART models at the moment. In the revision, I've currently included t5-small, and will include more models over the next few weeks.

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:51]: Do they exist as well in transformers?

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [reply optimum/neuron/generation/utils.py:100]: Alright

File: tests/test_generate.py
Comment (michaelbenayoun) [reply tests/test_generate.py:58]: Where is it tested for GPT and BART?
Anyways, alright let's do that!

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [reply optimum/neuron/generation/utils.py:744]: @aashiqmuhamed want to commit this?

File: optimum/neuron/generation/utils.py
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:25]: This causes an import error.
Can you do:

```
from ..utils import is_torch_xla_available
if is_torch_xla_available():
    import torch_xla.core.xla_model
```
Basically we want to be able to import and test code on regular machines when we do not need all of this.

File: tests/test_generate.py
Comment (michaelbenayoun) [inline tests/test_generate.py:75]: We specify that it's a trainium test, it will be ran on a Trainium instance
```suggestion
@is_trainium_test
@pytest.mark.parametrize("model_name, use_cache", testdata)
def test_greedy_decoding(model_name, use_cache):
    os.environ['XLA_USE_BF16'] = '0'
    xla_neuron_samples_fp32 = _test_greedy_decoding(model_name=model_name, device="xla")
    os.environ['XLA_USE_BF16'] = '1'
    xla_neuron_samples_bf16 = _test_greedy_decoding(model_name=model_name, device="xla")

    cpu_samples = _test_greedy_decoding(model_name=model_name, device="cpu")

    assert np.array_equal(cpu_samples, xla_neuron_samples_fp32), "XLA Neuron FP32 output doesn't match CPU only output"
    assert np.array_equal(cpu_samples, xla_neuron_samples_bf16), "XLA Neuron bf16 output doesn't match CPU only output"
```

File: tests/test_generate.py
Comment (michaelbenayoun) [inline tests/test_generate.py:13]:
```suggestion
from optimum.neuron.testing_utils import is_trainium_test
```

**Review by michaelbenayoun — COMMENTED** [review]
Some tests are failing, left comments to fix that.

**Review by michaelbenayoun — COMMENTED** [review]
Some tests are failing, left comments to fix that.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:51]: Yes will import them as well.

File: tests/test_generate.py
Comment (aashiqmuhamed) [reply tests/test_generate.py:58]: GPT and BART will be committed after we merge this CR. We are currently fixing a few bugs,

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:744]: Yes, I'm committing without the `is_torch_tpu_available()` check, since we expect to run on Trainium by default.

File: optimum/neuron/generation/utils.py
Comment (aashiqmuhamed) [reply optimum/neuron/generation/utils.py:25]: Got it, is the optimum neuron library designed for both CPU and Trainium?

Comment (michaelbenayoun) [reply optimum/neuron/generation/utils.py]: Only Trainium but it is more to be able to run non-trainum dependent tests on regular machines. Without this the test will fail if we dont have `torch_xla` installed even though we want to test something unrelated.

**Review by michaelbenayoun — APPROVED** [review]
LGTM!
Will merge if the tests pass.

**Comment by michaelbenayoun** [comment]
The EC2 runners cannot be used because they need secrets from this repo, which you do not have on your fork...
We can skip them for now.

Can you run the following command please:
```
make style
```
This will fix the styling error you currently have.

**Pull request merged**
